### PR TITLE
Use AtomicU64 for counters and gauges

### PR DESCRIPTION
**Pull request by olix0r**
Currently, tacho takes a mutex lock for all counter increments and gauge
writes. This is very slow.

By using the rust-nightly `integer_atomics` feature, we can substantially
improve throughput (by ~55.7%):

Previously:
```
target/release/examples/multithread  41.86s user 0.20s system 99% cpu 42.469 total
```

Now:
```
target/release/examples/multithread  18.53s user 0.03s system 99% cpu 18.627 total
```

I think I'd prefer to take a rust-nightly dependancy than accept this
performance hit. At least for the time being.

### CODE REVIEW

**Comment by carllerche** [comment]
Are 32 bit CPUs a huge concern? You can conditionally define `AtomicU64` to be `AtomicUsize` on  64 bit CPUs and use a mutex on 32 bit CPUs.

File: src/lib.rs [hunk 1]
```diff
@@ -197,23 +201,16 @@ impl Gauge {
...
-        let mut gauges = self.gauges
-            .write()
-            .expect("failed to obtain write lock for gauge");
-        if let Some(mut curr) = gauges.get_mut(&self.key) {
-            *curr = v;
-            return;
-        }
-        gauges.insert(self.key.clone(), v);
+    pub fn set(&self, v: u64) {
+        self.gauge.store(v, Ordering::Relaxed);
```
Comment (carllerche) [inline src/lib.rs]: `Relaxed` orderings always raise an eye, but given that no memory is being synchronized and it is a metric, it is probably fine. This, of course, runs the risk of the value of the gauge not matching the true last value (probably doesn't matter).

**Comment by olix0r** [comment]
Going to discard this and resubmit in the context of several other changes

**Comment by stevej** [comment]
I don't know the answer to this but is shipping production software on a nightly rustc a good idea? Do others do it?

**Comment by olix0r** [comment]
@stevej #18 redoes this without a nightly dependency. that said, our CI, etc is nightly. And linkerd-tcp is still marked experimental ;)

**Pull request closed without merging**